### PR TITLE
fix: prefix relation id columns contained in embedded entities (#6977)

### PR DESCRIPTION
**Pull request by nebkat**
Searches embedded entity columns for relation ID column if relation column is in embedded entity. If not found, creates new relation ID with embedded metadata set to match the relation column.

Fixes: #2254
Fixes: #3132
Fixes: #3226
Fixes: #6977

### Description of change

See #6977. 

Currently the relation column's `entityMetadata` is searched for columns matching the join column, rather than `embeddedMetadata` (when present).

https://github.com/typeorm/typeorm/blob/a5eb946117a18d94c0157188b6a39542c8d50756/src/metadata-builder/RelationJoinColumnBuilder.ts#L136.

This changes that behavior to match 

https://github.com/typeorm/typeorm/blob/a5eb946117a18d94c0157188b6a39542c8d50756/src/metadata-builder/EntityMetadataBuilder.ts#L152-L156

### Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] `npm run lint` passes with this change
- [x] `npm run test` passes with this change
- [x] This pull request links relevant issues as `Fixes #0000`
- [x] There are new or updated unit tests validating the change
- [x] Documentation has been updated to reflect this change
- [x] The new commits follow conventions explained in [CONTRIBUTING.md](https://github.com/typeorm/typeorm/blob/master/CONTRIBUTING.md)

### CODE REVIEW

**Comment by imnotjames** [comment]
Looks like this change causes the oracle tests to fail?

**Comment by nebkat** [comment]
Oracle issue with max identifier length being 30 characters, looks like it never occurred in the past because the column names always excluded the embedded prefixes. Fixed by shortening a prefix in the edge cases with long identifiers.

File: src/metadata-builder/RelationJoinColumnBuilder.ts [hunk 1]
```diff
@@ -133,11 +133,13 @@ export class RelationJoinColumnBuilder {
             });
             const joinColumnName = joinColumnMetadataArg ? joinColumnMetadataArg.name : this.connection.namingStrategy.joinColumnName(relation.propertyName, referencedColumn.propertyName);
 
-            let relationalColumn = relation.entityMetadata.ownColumns.find(column => column.databaseName === joinColumnName);
+            const relationalColumns = relation.embeddedMetadata ? relation.embeddedMetadata.columns : relation.entityMetadata.ownColumns;
```
Comment (pleerock) [inline src/metadata-builder/RelationJoinColumnBuilder.ts:136]: what if there is an embedded, but we point to a relation that not in the embedded?

File: test/functional/entity-metadata/entity/Counters.ts [hunk 2]
```diff
@@ -18,7 +18,7 @@ export class Counters {
     @Column()
     favorites: number;
 
-    @Column(() => Subcounters)
+    @Column(() => Subcounters, {prefix: "sub"})
```
Comment (pleerock) [inline test/functional/entity-metadata/entity/Counters.ts:21]: Why did you change exist test? You need to create a separate test for a new use case.

Comment (nebkat) [reply test/functional/entity-metadata/entity/Counters.ts:21]: The columns inside embedded entities were being incorrectly generated previously. The correct column name is `countersSubcountersWatchedUserId`, which is 32 characters long, and too long for Oracle DB which can only take 30 characters. Before this fix, the column name would have been `watchedUserId` or something, so there was no problem, but the wrong column name was going unnoticed.

This test only checks the ORM side of things anyway, so the exact name in database was irrelevant, but I can change the variable name from `subcounters` to `sub` directly if you'd prefer.  I have already created a new test case for the issue #6977.

File: src/metadata-builder/RelationJoinColumnBuilder.ts (hunk 1 above)
Comment (nebkat) [reply src/metadata-builder/RelationJoinColumnBuilder.ts:136]: I'm not sure what you mean, I think `relation.embeddedMetadata` is only set if `relation` is a column within an embedded entity. It was somewhat confusing initially but `relation` refers only to the foreign key column, not the target relation entity. 

This particular section is what searches for an existing relation id column, and if the relation is located inside an embedded then it only makes sense to search for the relation id column also in the embedded rather than in the base of the entity.

File: test/functional/entity-metadata/entity/Counters.ts (hunk 2 above)
Comment (pleerock) [reply test/functional/entity-metadata/entity/Counters.ts:21]: the problem is, if we merge these changes, we introduce a breaking change. Right now Oracle users have same case working (even if column name is invalid), but if we merge these changes their codebases won't be working at all. We can't tell them "hey guys just monkey patch it using "prefix" like we did it on our tests". Proper solution is to fix related Oracle issue.

**Comment by pleerock** [comment]
@imnotjames it looks like this will introduce a serious breaking change for the users who have same use case. Right now what we have is a bug, this is true. If we merge these changes bug will be fixed, but it will also lead to schema re-creation and data loss for users. We need to decide how much acceptable it is to merge these changes not into the `next`.

File: src/metadata-builder/RelationJoinColumnBuilder.ts (hunk 1 above)
Comment (pleerock) [reply src/metadata-builder/RelationJoinColumnBuilder.ts:136]: okay, thank you for explanation

File: test/functional/entity-metadata/entity/Counters.ts (hunk 2 above)
Comment (nebkat) [reply test/functional/entity-metadata/entity/Counters.ts:21]: For this test anyway there is no solution for Oracle unless we plan to truncate column names automatically? Could potentially introduce a more informative error during column generation if target DB is Oracle since the DB error is very vague.

Comment (pleerock) [reply test/functional/entity-metadata/entity/Counters.ts:21]: we already do truncate in [some other places](https://github.com/typeorm/typeorm/blob/494866a999606e4acf8275aaae14099b0c1a0be1/src/find-options/FindOptionsUtils.ts#L229) for Oracle

**Comment by nebkat** [comment]
I see the difficulty of maintaining the project now 🙂 Would it be worth introducing a temporary option property to relations that applies this fix until next major release? Definitely shouldn't break existing setups, but unless the next release is coming soon it might be frustrating for users and will also lead to more broken columns being initiated in new projects. In my case I have multiple userId references and it fails because of duplicate column so I can't even keep the ORM side correct.

`@ManyToOne(() => UserEntity, {fixEmbeddedRelationColumn: true})`?

**Comment by pleerock** [comment]
if we go this way, but better to make same a connection-level option

**Comment by nebkat** [comment]
Sounds good, if you think this is best I can implement it. I propose maybe a field in `BaseConnectionOptions` e.g. `targetVersion: "0.3.0"` which defaults to the current release, string comparison as a condition for the breaking changes, along with a strong warning for users that things are likely to break if they target future versions. Once that version is released the old code gets removed, but hopefully additional feedback can be gained from users willing to test new features.

File: src/metadata-builder/RelationJoinColumnBuilder.ts (hunk 1 above)
Comment (nebkat) [reply src/metadata-builder/RelationJoinColumnBuilder.ts:136]: I've had a second look at this, and in theory we could allow relation IDs to not be in the embedded entity together with the relation, but I'm not sure whether that would cause additional confusion.

```typescript
@Entity()
class Other {
    @PrimaryColumn() id: number;
}
@Entity()
class User {
    @Column(() => Embed) embed: Embed;

    @Column() embedOther1Id: number;
    @ManyToOne(() => Other) embedOther2: Other;
}
class Embed {
    @ManyToOne(() => Other) other1: Other;
    @Column() other2Id: number;
}
```

Both relations here could match up correctly if we want, and if they are not matched up they will lead to a duplicate column error anyway. MongoDB is an exception since it maintains the embedded objects in DB, so both relation/ID have to be in the same entity.

**Comment by pleerock** [comment]
I think it's better just to call them "enableNextFeatures", don't think we'll really try to support more than one next breaking version.

File: test/functional/entity-metadata/entity/Counters.ts (hunk 2 above)
Comment (nebkat) [reply test/functional/entity-metadata/entity/Counters.ts:21]: Might be something worth looking at with naming strategies for the future, gives the user the option for more meaningful change than simple truncation.

**Comment by nebkat** [comment]
Since latest TypeScript is merged the next release should be considered breaking, can the `enableNextFeatures` flag be removed and have the change applied by default?
